shell/shell.go

```diff
@@ -48,47 +49,22 @@ func Parse(s string) ([]Command, []int, error) {
 		return nil, nil, nil
 	}
 	// First syntax pass: Eliminate anything verboten.
-	// Be very conservative for now.
+	// Be very conservative for now, using an allowlist.
 	syntax.Walk(f, func(n syntax.Node) bool {
 		switch n := n.(type) {
-		case *syntax.IfClause:
-			err = fmt.Errorf("if statements are not supported")
-		case *syntax.Block:
-			err = fmt.Errorf("shell blocks are not supported")
-		case *syntax.Redirect:
-			err = fmt.Errorf("redirects are not supported")
-		case *syntax.CallExpr:
-			if len(n.Assigns) > 0 {
-				err = fmt.Errorf("variables are not supported")
-			}
-		case *syntax.ProcSubst:
-			err = fmt.Errorf("process substitution is not supported")
-		case *syntax.DblQuoted:
-			if n.Dollar {
-				err = fmt.Errorf("dollar in double quotes is not supported")
-			}
-		case *syntax.Subshell:
-			err = fmt.Errorf("subshells are not supported")
-		case *syntax.ParamExp:
-			err = fmt.Errorf("parameter expansion is not supported")
-		case *syntax.CmdSubst:
-			err = fmt.Errorf("command substitution is not supported")
-		case *syntax.ArithmExp:
-			err = fmt.Errorf("arithmetic expansion is not supported")
-		case *syntax.ExtGlob:
-			err = fmt.Errorf("extended globs are not supported")
-		case *syntax.Comment:
-			// comments make it harder to handle trailing pipes
-			err = fmt.Errorf("comments are not supported")
-		case *syntax.BinaryArithm:
-			err = fmt.Errorf("binary arithmetic is not supported")
+		case nil, *syntax.File, *syntax.CallExpr, *syntax.Word,
+			*syntax.Lit, *syntax.SglQuoted, *syntax.DblQuoted:
 		case *syntax.BinaryCmd:
-			switch n.Op {
-			case syntax.AndStmt:
-				err = fmt.Errorf("&& is not supported")
-			case syntax.OrStmt:
-				err = fmt.Errorf("|| is not supported")
+			// TODO: consider supporting |& (syntax.PipeAll) to pipe stderr as well
```

Comment:
Right now it always pipes stderr. It turned out to be a better default, so that error messages on malformed inputs just worked. So if we supported `|&`, would it...toggle it back off? Or maybe do something fancier where we display stderr by default but don't incorporate the output into subsequent commands without `|&`?

I'm not sure what the best behavior is; that's why it is omitted for now, not (just) laziness.

Reply:
ah gotcha, makes sense. I think it's perfectly fine to implement a slightly different shell for your purposes here, as long as it's documented and the error is helpful :)